Update README (#3)
sangria-graphql/sangria-federated: README.md

![Sangria Logo](https://sangria-graphql.github.io/assets/img/sangria-logo.svg)

# sangria-federated

![Continuous Integration](https://github.com/sangria-graphql/sangria-federated/workflows/Continuous%20Integration/badge.svg)
[![Maven Central](https://maven-badges.herokuapp.com/maven-central/org.sangria-graphql/sangria-federated_2.13/badge.svg)](https://maven-badges.herokuapp.com/maven-central/org.sangria-graphql/sangria-federated_2.13)
[![License](http://img.shields.io/:license-Apache%202-brightgreen.svg)](http://www.apache.org/licenses/LICENSE-2.0.txt)
[![Scaladocs](https://www.javadoc.io/badge/org.sangria-graphql/sangria-federated_2.13.svg?label=docs)](https://www.javadoc.io/doc/org.sangria-graphql/sangria-federated_2.13)
[![Join the chat at https://gitter.im/sangria-graphql/sangria](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/sangria-graphql/sangria?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

> **sangria-federated** is a library that allows sangria users to implement services that adhere to
> [Apollo's Federation Specification](https://www.apollographql.com/docs/federation/federation-spec/), and can be used as
> part of a federated data graph.

Reviewer: ```suggestion
**sangria-federated** is a library that allows sangria users to implement services that adhere to [Apollo's Federation Specification](https://www.apollographql.com/docs/federation/federation-spec/), and can be used as part of a federated data graph.
```

SBT Configuration:

```scala

> libraryDependencies += "org.sangria-graphql" %% "sangria-federated" % "latest.release"

Reviewer: This is the convention we use on other libraries:
```suggestion
libraryDependencies += "org.sangria-graphql" %% "sangria-federated" % "<latest version>"
```

Author: but this doesn't compile, while the other does

Reviewer: `libraryDependencies += "org.sangria-graphql" %% "sangria-federated" % "latest.release"` compiles. But is it correct?

The goal is only to display a template that people can copy/paste and adapt.

Author: alright 👍 I changed it

```

## How does it work?

> The library adds [Apollo's Federation Specification](https://www.apollographql.com/docs/federation/federation-spec/)
> on top of the provided sangria graphql schema.

Reviewer: ```suggestion
The library adds [Apollo's Federation Specification](https://www.apollographql.com/docs/federation/federation-spec/) on top of the provided sangria graphql schema.
```

To make it possible to use `_Any` as a scalar, the library upgrades the used marshaller.

## How to use it?

> To be able to communicate with [Apollo's federation gateway](https://www.apollographql.com/docs/federation/gateway/), 
> the graphql sangria service should be using both the federated schema and unmarshaller.

Reviewer: ```suggestion
To be able to communicate with [Apollo's federation gateway](https://www.apollographql.com/docs/federation/gateway/), the graphql sangria service should be using both the federated schema and unmarshaller.
```

    
      val stateResolver = EntityResolver[StateService, Json, State, StateArg](
        __typeName = "State",
        ev = State.decoder,
        arg =>  env.getState(arg.id))
    
      val stateSchema =
        ObjectType(
          "State",
          fields[Unit, State](
            Field(
              "id", IntType,
              resolve = _.value.id),
            Field(
              "key", StringType,
              resolve = _.value.key),
            Field(
              "value", StringType,
              resolve = _.value.value))
        ).copy(astDirectives = Vector(federation.Directives.Key("id")))
    }
    ```
  
    The entity resolver implements:
    - the deserialization of the fields in `_Any` object to the EntityArg.
    - how to fetch the EntityArg to get the proper Entity (in our case State).
    
    As for the query type, let's suppose the schema below:
    ```scala
    import sangria.schema._
    
    object StateAPI {
    
    val Query = ObjectType(
      "Query",
      fields[StateService, Unit](
        Field(
          name = "states", 
          fieldType = ListType(State.stateSchema), 
          resolve = _.ctx.getStates)))
    }
    ```
  

>     Now in the definition of the GraphQL server, we federate the Query type and the unmarshaller while supplying the
>     entity resolvers. Then, we use both the federated schema and unmarshaller as arguments for the server.

Reviewer: ```suggestion
    Now, in the definition of the GraphQL server, we federate the Query type and the unmarshaller while supplying the entity resolvers. Then, we use both the federated schema and unmarshaller as arguments for the server.
```

        import scala.concurrent.ExecutionContext.Implicits.global
        
        def exec(
          schema:        Schema[A, Unit],
          userContext:   F[A],
          query:         Document,
          operationName: Option[String],
          variables:     Json): F[Either[Json, Json]] = userContext.flatMap { ctx =>
          
            F.async { (cb: Either[Throwable, Json] => Unit) =>
              Executor.execute(
                schema           = schema,
                queryAst         = query,
                userContext      = ctx,
                variables        = variables,
                operationName    = operationName,
                exceptionHandler = ExceptionHandler {
                  case (_, e) ⇒ HandledException(e.getMessage)
                }
              ).onComplete {
                case Success(value) => cb(Right(value))
                case Failure(error) => cb(Left(error))
              }
            }
          }.attempt.flatMap {
            case Right(json)               => F.pure(json.asRight)
            case Left(err: WithViolations) => ???
            case Left(err)                 => ???
          }
      }
    }
    ```
  

> - The review service defines the review type, which has a reference to the state type. And, for each entity referenced
>   by another service, a [stub type](https://www.apollographql.com/docs/federation/entities/#referencing) should be 
>   created (containing just the minimal information that will allow to reference the entity).

Reviewer: ```suggestion
- The review service defines the review type, which has a reference to the state type. And, for each entity referenced by another service, a [stub type](https://www.apollographql.com/docs/federation/entities/#referencing) should be created (containing just the minimal information that will allow to reference the entity).
```

            resolve = _.value.state)))
  }
  
  case class State(id: Int)

  object State {
    
    import sangria.federation.Directives._
  
    val stateSchema =
      ObjectType(
        "State",
        fields[Unit, State](
          Field[Unit, State, Int, Int](
            name = "id",
            fieldType = IntType,
            resolve = _.value.id).copy(astDirectives = Vector(External)))
      ).copy(astDirectives = Vector(Key("id"), Extends))
  }
  ```
  
  In the end, the same code used to federate the state service is used to federate the review service.

- The sangria GraphQL services endpoints can now be configured in the ```serviceList``` of [Apollo's Gatewqay](
  https://www.apollographql.com/docs/federation/gateway/#setup) as follows:
    ```
    const gateway = new ApolloGateway({
      serviceList: [
        { name: 'states', url: 'http://localhost:9081/api/graphql'},
        { name: 'reviews', url: 'http://localhost:9082/api/graphql'}
      ],
      debug: true
    })
    ```
  
All the code of the example is available [here](./example).

## Caution 🚨🚨

- **This is a technology preview and should not be used in a production environment.**

> - The library upgrades the marshaller too, by making map values scalars (e.g. json objects as scalars). This can lead to
>   security issues as discussed [here](
>   http://www.petecorey.com/blog/2017/06/12/graphql-nosql-injection-through-json-types/).

Reviewer: ```suggestion
- The library upgrades the marshaller too, by making map values scalars (e.g. json objects as scalars). This can lead to security issues as discussed [here](http://www.petecorey.com/blog/2017/06/12/graphql-nosql-injection-through-json-types/).
```

# sangria-federated

![Continuous Integration](https://github.com/sangria-graphql/sangria-federated/workflows/Continuous%20Integration/badge.svg)
[![Maven Central](https://maven-badges.herokuapp.com/maven-central/org.sangria-graphql/sangria-federated_2.13/badge.svg)](https://maven-badges.herokuapp.com/maven-central/org.sangria-graphql/sangria-federated_2.13)
[![License](http://img.shields.io/:license-Apache%202-brightgreen.svg)](http://www.apache.org/licenses/LICENSE-2.0.txt)
[![Scaladocs](https://www.javadoc.io/badge/org.sangria-graphql/sangria-federated_2.13.svg?label=docs)](https://www.javadoc.io/doc/org.sangria-graphql/sangria-federated_2.13)
[![Join the chat at https://gitter.im/sangria-graphql/sangria](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/sangria-graphql/sangria?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

**sangria-federated** is a library that allows sangria users to implement services that adhere to
[Apollo's Federation Specification](https://www.apollographql.com/docs/federation/federation-spec/),
and can be used as part of a federated data graph.

SBT Configuration:

```scala

> libraryDependencies += "org.sangria-graphql" %% "sangria-federated" % "0.0.1"

Reviewer: I don't want to change the README each time we release a new version.
```suggestion
libraryDependencies += "org.sangria-graphql" %% "sangria-federated" % "<last version>"
```

Author: :) I discovered that if I put :
```scala
libraryDependencies += "org.sangria-graphql" %% "sangria-federated" % "latest.release"
```
Ivy will automatically get the latest release (doesn't work when I use <last version>), this way the user doesn't have to change the code.

```

## How does it work?

> The library adds [Apollo's Federation Specification](https://www.apollographql.com/docs/federation/federation-spec/)
> on top of the provided sangria graphql schema. And to make it possible to use *_Any* as a scalar,
> the library upgrades the marshaller used in the user application too.

Reviewer: ```suggestion
The library adds [Apollo's Federation Specification](https://www.apollographql.com/docs/federation/federation-spec/)
on top of the provided sangria graphql schema.

To make it possible to use *_Any* as a scalar, the library upgrades the used marshaller.
```

## How to use it?

> To be able to communicate with [Apollo's federation gateway](https://www.apollographql.com/docs/federation/gateway/),
> the graphql sangria service should be using both the federated schema and unmarshaller.

Reviewer: ```suggestion
To be able to communicate with [Apollo's federation gateway](https://www.apollographql.com/docs/federation/gateway/), the graphql sangria service should be using both the federated schema and unmarshaller.
```

As an example, let's consider an application using circe with a state and review service. 

> - The state service defines the state entity, annotated with ```@key("id")```. And for each entity, 
>   we need to define an entity resolver ([reference resolver](https://www.apollographql.com/docs/federation/entities/#resolving))
>   , see code below:

Reviewer: Please let the browser inserting new lines.

```suggestion
- The state service defines the state entity, annotated with ```@key("id")```. For each entity, we need to define an entity resolver ([reference resolver](https://www.apollographql.com/docs/federation/entities/#resolving)) as below:
```

Author: I have a 120 line limit in Intellij, this way I avoid long lines and I don't have to scroll. If it's okay for you I will try to break only lines that are surpassing this limit 🙏

    ```scala
    import sangria.federation.Decoder
    import io.circe.Json,  io.circe.generic.semiauto._
    import sangria.schema._
    
    case class State(
      id: Int,
      key: String,
      value: String)
    
    object State {
    
      case class StateArg(id: Int)
    
      implicit val decoder: Decoder[Json, StateArg] = deriveDecoder[StateArg].decodeJson(_)
    
      val stateResolver = EntityResolver[StateService, Json, State, StateArg](
        __typeName = "State",
        ev = State.decoder,
        arg =>  env.getState(arg.id))
    

>       implicit val stateSchema =

Reviewer: Does it need to be implicit?

Author: no, you are right 🙏 it does not need to be, I changed it in the last commit.

        ObjectType(
          "State",
          fields[Unit, State](
            Field(
              "id", IntType,
              resolve = _.value.id),
            Field(
              "key", StringType,
              resolve = _.value.key),
            Field(
              "value", StringType,
              resolve = _.value.value))

>         ).copy(astDirectives = Vector(federation.Directives.Key("id")))

Reviewer: Can we define the directive on the `id` Field itself?

Other option: maybe we can use the macro based approach in the example.

Author: No, if you take a look at how [apollo](https://www.apollographql.com/docs/federation/entities/#defining) uses it, the directive is used on the type (but there are others which are used on the [field](https://github.com/sangria-graphql/sangria-federated/blame/1001268c6a57d10b08fd73f9768b369074deb74c/README.md#L196)). And, yes I also agree on using macros, using annotations is much user friendly. In fact, I already started working on it a few weeks ago, but it still needs some work 😞

    }
    ```
    

>     The entity resolver, must include how to deserialize the fields in _Any object to the EntityArg.
>     And, how to use the EntityArg to get the proper Entity (in our case State).

Reviewer: ```suggestion
    The entity resolver implements:
    - the deserialization of  the fields in `_Any` object to the EntityArg.
    - how to fetch the EntityArg to get the proper Entity (in our case State).
```

  
    As for the query type, let's suppose the schema below:
    ```scala
    import sangria.schema._
    
    object StateAPI {
    
    val Query = ObjectType(
      "Query",
      fields[StateService, Unit](
        Field(
          name = "states", 
          fieldType = ListType(State.stateSchema), 
          resolve = _.ctx.getStates)))
    }
    ```
    
    Now in the definition of the GraphQL server, we federate the Query type and the unmarshaller 
    while supplying the entity resolvers. Then, we use both the federated schema and unmarshaller as
    arguments for the server.
    ```scala
    def graphQL[F[_]: Effect]: GraphQL[F] = {
      val (schema, um) = federation.Federation.federate[StateService, Json](
        Schema(StateAPI.Query),
        sangria.marshalling.circe.CirceInputUnmarshaller,
        stateResolver)
    
      GraphQL(
        schema,
        env.pure[F])(implicitly[Effect[F]], um)
    }
    ```
  
    And, the GraphQL server should use the provided schema and unmarshaller as arguments for the
    sangria executor:
    ```scala
    import cats.effect._

>     import cats.implicits._

Reviewer: We should stick with `Future`  in the example.

Author: If I change IO to Future, it won't really change the semantic of the tuto (only used in the query execution), and it will only cost time to implement it again :/ If you still think I should stick to Future then I can open another PR for it

Reviewer: Yes please stick to Future in the Readme. A lot of people are using sangria because it has no dependencies.

Author: Alright :) I noted it I will do another PR for it, to change both the example in README, after this one get merged 🙏

  should be created (containing just the minimal information that will allow to reference
  the entity).
  ```scala
  import sangria.schema._
  
  case class Review(
    id: Int,
    key: Option[String] = None,
    state: State)
    
  object Review {

    implicit val reviewSchema =
      ObjectType(
        "Review",
        fields[Unit, Review](
          Field(
            "id", IntType,
            resolve = _.value.id),
          Field(
            "key", OptionType(StringType),
            resolve = _.value.key),
          Field(
            "state", State.stateSchema,
            resolve = _.value.state)))
  }
  
  case class State(id: Int)

  object State {
    
    import sangria.federation.Directives._
  
    implicit val stateSchema =
      ObjectType(
        "State",
        fields[Unit, State](
          Field[Unit, State, Int, Int](
            name = "id",
            fieldType = IntType,
            resolve = _.value.id).copy(astDirectives = Vector(External)))
      ).copy(astDirectives = Vector(Key("id"), Extends))
  }
  ```
  
  In the end, the same code used to federate the state service is used to federate the review

>   service.

Reviewer: Is it about how to the state service and the review service expose the GraphQL schema?

Author: yes, I am referring to this [part](https://github.com/sangria-graphql/sangria-federated/pull/3/files/1001268c6a57d10b08fd73f9768b369074deb74c#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R96)

  

- The sangria GraphQL services endpoints can now be configured in the ```serviceList``` of
  [Apollo's Gatewqay](https://www.apollographql.com/docs/federation/gateway/#setup) as follows:
    ```
    const gateway = new ApolloGateway({
      serviceList: [
        { name: 'states', url: 'http://localhost:9081/api/graphql'},
        { name: 'reviews', url: 'http://localhost:9082/api/graphql'}
      ],
      debug: true
    })
    ```

All the code of the example is available [here](./example).

## Caution 🚨🚨

- **This is a technology preview and should not be used in a production environment.**

> - The library upgrades the marshaller too, by making map values scalars (e.g. json objects as 
> scalars). This if not treated with caution, can lead to security issues as discussed [here](http://www.petecorey.com/blog/2017/06/12/graphql-nosql-injection-through-json-types/).

Reviewer: ```suggestion
- The library upgrades the marshaller by making map values scalars (e.g. json objects as  scalars). This can lead to security issues as discussed [here](http://www.petecorey.com/blog/2017/06/12/graphql-nosql-injection-through-json-types/).
```

## Contribute

> Contributions are warmly desired 🤗; please follow the standard process of forking the repo and making
> PRs 🤓

Reviewer: ```suggestion
Contributions are warmly desired 🤗. Please follow the standard process of forking the repo and making PRs 🤓
```

          Field(
            "key", OptionType(StringType),
            resolve = _.value.key),
          Field(
            "state", State.stateSchema,
            resolve = _.value.state)))
  }
  
  case class State(id: Int)

  object State {
    
    import sangria.federation.Directives._
  
    implicit val stateSchema =
      ObjectType(
        "State",
        fields[Unit, State](
          Field[Unit, State, Int, Int](
            name = "id",
            fieldType = IntType,
            resolve = _.value.id).copy(astDirectives = Vector(External)))
      ).copy(astDirectives = Vector(Key("id"), Extends))
  }
  ```
  
  In the end, the same code used to federate the state service is used to federate the review
  service.

- The sangria GraphQL services endpoints can now be configured in the ```serviceList``` of
  [Apollo's Gatewqay](https://www.apollographql.com/docs/federation/gateway/#setup) as follows:
    ```
    const gateway = new ApolloGateway({
      serviceList: [
        { name: 'states', url: 'http://localhost:9081/api/graphql'},
        { name: 'reviews', url: 'http://localhost:9082/api/graphql'}
      ],
      debug: true
    })
    ```

> All the code of the example is available here.

Author: > If we have a test showing how to use it, we could simply add a link from the README to the test.
That would ensure that the test stays up-to-date.

I agree with you on this too, this is why I wrote this phrase at the end. l was hoping to update it when I push the next example code PR. But, I am not against postponing it :)
